Report event for the cases when probe returned Unknown result

jralmaraz opened the pull request:
This fix aims to expose in logs the cases when the Unknown result can be returned by the probe.

The attempt of this fix is to:

- Ensure that when a probe result is Unknown, a log entry is made;
- That an event is recorded and exposed when a probe result is Unknown.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes: 
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #116026 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Implemented logging and event recording for probe results with an `Unknown` status in the kubelet's prober module. This helps in better diagnosing and monitoring cases where container probes return an `Unknown` result, improving the observability and reliability of health checks.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link> 
- [Other doc]: <link>
-->
```docs
- [Usage]: [Kubernetes Probes Documentation](https://github.com/kubernetes/website/blob/cfc5ea3a3b87b7895943ecca4656cf832f45c8c7/content/en/docs/concepts/workloads/pods/pod-lifecycle.md#container-probes)
```

jralmaraz committed:
Report event for the cases when probe returned Unknown result

keep log success

jralmaraz commented:
/check-cla

jralmaraz commented:
/check-cla

jralmaraz commented:
/test all

Octopusjust commented:
The issue number should be filled in under `"Which issue(s) this PR fixes:"`

jralmaraz commented:
> The issue number should be filled in under `"Which issue(s) this PR fixes:"`

Updated, thank you!

jralmaraz commented:
hi @Octopusjust , I see the tests failing for different things that this PR doesn't touch.

Is there anything I can do on my end to workaround them ?

Thanks!

[pull-kubernetes-e2e-gce](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/125901/pull-kubernetes-e2e-gce/1812683309130452992)

[pull-kubernetes-node-e2e-containerd](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/125901/pull-kubernetes-node-e2e-containerd/1812683309780570112)

jralmaraz commented:
/retest

bart0sh commented:
@jralmaraz Please, look at the previous attempts to fix https://github.com/kubernetes/kubernetes/issues/116026 if you haven't done it yet, thanks! There could be some hints there that you can use.
- https://github.com/kubernetes/kubernetes/pull/116030
- https://github.com/kubernetes/kubernetes/pull/116077
- https://github.com/kubernetes/kubernetes/pull/117054
- https://github.com/kubernetes/kubernetes/pull/118610
- https://github.com/kubernetes/kubernetes/pull/125373

BTW, the latter one is still open. To have yours approved it should probably be better than previous ones. Is it ;) ?

bart0sh commented:
/kind bug
/cc

@jralmaraz as this is user-visible change, please update PR description and provide a release note, thanks.

bart0sh commented:
/triage accepted
/priority backlog

jralmaraz commented:
> @jralmaraz Please, look at the previous attempts to fix #116026 if you haven't done it yet, thanks! There could be some hints there that you can use.
> 
> * [The event message will include the probe type and the unknown status #116030](https://github.com/kubernetes/kubernetes/pull/116030)
> * [Prober: Implement logging and event for unknown probe result  #116077](https://github.com/kubernetes/kubernetes/pull/116077)
> * [add probe.Unknown message #117054](https://github.com/kubernetes/kubernetes/pull/117054)
> * [Adding an event for unknown results when there is no error and added … #118610](https://github.com/kubernetes/kubernetes/pull/118610)
> * [Add events and logs for when the prober returns an unknown result  #125373](https://github.com/kubernetes/kubernetes/pull/125373)
> 
> BTW, the latter one is still open. To have yours approved it should probably be better than previous ones. Is it ;) ?

hi @bart0sh , thanks for your review and inputs. 

The difference I can see from the opened PR is that I've added a test case - the core fix is very similar, but I have also gathered for a few other conditions/cases (HTTP probe request to POD and TCP request to resolve container port)  to return those to log for unknown prober events.

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +333,40 @@ func TestNewExecInContainer(t *testing.T) {
+	if prober == nil {
+		t.Error("Expected a non-nil prober")
+	}
+	// Add more assertions as needed to validate the initialization of prober
```
@jralmaraz Please, consider testing events. Please, use previous PRs, e.g. https://github.com/kubernetes/kubernetes/pull/116030/files#diff-a4e1cdbcf3b09c95dcabb07e531db97347f01c1d5abc151d867e9221a1caa47cR338-R409

jralmaraz replied:
I've reviewed the other PRs and added testing of events as well, let me know if it look ok! Thanks a lot!

bart0sh commented:
> The difference I can see from the opened PR is that I've added a test case

Please, look at the review comments from previous PRs and make sure they're all addressed or not relevant for your changes, thanks.

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +335,141 @@ func TestNewExecInContainer(t *testing.T) {
+}
+
+func TestExecInContainer_Start(t *testing.T) {
+	ctx := context.Background()
```
```suggestion
	ctx := ktesting.Init(t)
```

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +335,141 @@ func TestNewExecInContainer(t *testing.T) {
+
+	exec := prober.newExecInContainer(ctx, container, containerID, cmd, 0)
+	err := exec.Start()
+	if err != nil {
+		t.Errorf("Expected no error, got %v", err)
+	}
```
please, use `require.Noerror(err)` or `assert.NoError(err)` in this and similar cases.

bart0sh commented:
@jralmaraz Overall it looks good to me. Please, address minor things from review comments and squash commits into one, thanks.

jralmaraz commented:
/retest

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +333,138 @@ func TestNewExecInContainer(t *testing.T) {
+func TestRecordContainerEventUnknownStatus(t *testing.T) {
+
+	if err := v1.AddToScheme(legacyscheme.Scheme); err != nil {
+		t.Errorf("failed to add v1 to scheme: %v", err)
```
Can you use either `assert.NoError` or `require.NoError` here?

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -237,24 +241,20 @@ func TestProbe(t *testing.T) {
-					t.Errorf("[%s] Didn't expect probe error but got: %v", testID, err)
-					continue
-				}
+				require.NoError(t, err, "[%s] Didn't expect probe error but got: %v", testID, err)
```
As far as I remember `require.NoError` includes `err` into its default message, so you don't need to do it.

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +333,138 @@ func TestNewExecInContainer(t *testing.T) {
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.got != tt.expect {
```
Please use either `assert.Equal` or `require.Equal` for more readability and consistency, thanks.

bart0sh replied:
I expected that test cases would include some parameters and tested function (`newProber`) would be called for every test case. Calling it only once makes me think that it would be easier to do it without `tests`, e.g.
```
t.Run(tt.name, func(t *testing.T) {
        runner := &containertest.FakeContainerCommandRunner{}
	recorder := &record.FakeRecorder{}
	
       prober := newProber(runner, recorder)
       assert.NotNil(prober)
       assert.Equal(runner, prober.runner)
       ....
}
```

jralmaraz replied:
Hopefully this one has been resolved in the most recent changes.

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +333,138 @@ func TestNewExecInContainer(t *testing.T) {
+
+		// Receive events from the channel
+		for i := 0; i < 2; i++ {
+			events = append(events, <-recorder.Events)
```
Would this block the test if second even is not received for some reason?

jralmaraz replied:
Indeed - tried to create a buffered channel to avoid blocking. Is that a good approach ?

bart0sh replied:
I don't know. Have you tested if it helps? How similar tests work with `FakeRecorder`?

bart0sh replied:
@jralmaraz This one is still not addressed. Have you looked at other tests that use FakeRecorder? Do they use the same approach?

jralmaraz replied:
Hi @bart0sh - sorry for leaving this one behind. I have changed the implementation to created the `FakeRecorder` with the buffer size based on the length of the expected events in the test cases. Is that an ok approach ? 

I checked the [cronjob_controllerv2_test](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controllerv2_test.go#L1321) , [daemon_controller_test](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemon_controller_test.go#L331) and a few others, but still trying to learn how to best use it.

bart0sh replied:
> Is that an ok approach ?

I'm not sure it's ok. I'd at least check if amount of events is expected before reading them.

BTW, you can comment out one of of the `pb.recordContainerEvent` calls to see that making a bigger buffer doesn't make reading from an empty channel successful.

jralmaraz replied:
Currently commenting it was timing out, so I added a check to fail before, but still working on it and reviewing before pushing again.

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +333,138 @@ func TestNewExecInContainer(t *testing.T) {
+		}
+
+		for i, event := range events {
+			if event != tc.expected[i] {
```
please use `assert.Equal`.

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -112,6 +112,9 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
 		klog.V(3).InfoS("Probe succeeded with a warning", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "output", output)
+	} else if result == probe.Unknown {
+		klog.V(3).InfoS("Probe resulted in unknown state", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
+		pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerProbeWarning, "%s probe resulted in unknown state: %s", probeType, output)
```
Would it be more readable to use `switch` instead of bunch of `if` and `else if` statements?

jralmaraz replied:
I've used `switch` before in this PR, but I think the `lint` build failed: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/125901/pull-kubernetes-linter-hints/1821041892498870272

Any pointer to get `linter` working in this case ?

jralmaraz replied:
Hopefully that one has been resolved with latest changes.

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -98,7 +98,7 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
 
 	result, output, err := pb.runProbeWithRetries(ctx, probeType, probeSpec, pod, status, container, containerID, maxProbeRetries)
-	if err != nil || (result != probe.Success && result != probe.Warning) {
+	if err != nil || (result != probe.Success && result != probe.Warning) && result != probe.Unknown {
```
Would it simplify the code if we check for `probe.Success`, `probe.Failure`, `probe.Warning` and `probe.Unknown` separately?

jralmaraz commented on pkg/kubelet/prober/prober.go:
```diff
@@ -144,6 +147,9 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
 		req, err := httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
 		if err != nil {
+			// Log and record event for Unknown result
+			klog.InfoS("HTTP-Probe failed to create request", "error", err)
```
Thanks @toVersus  - tried to fix it in other places too.

jralmaraz commented on pkg/kubelet/prober/prober.go:
```diff
@@ -98,7 +98,7 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
 
 	result, output, err := pb.runProbeWithRetries(ctx, probeType, probeSpec, pod, status, container, containerID, maxProbeRetries)
-	if err != nil || (result != probe.Success && result != probe.Warning) {
+	if err != nil || (result != probe.Success && result != probe.Warning) && result != probe.Unknown {
```
You are absolutely right ! So sorry, I must have toggled during some local tests. I've pushed some changes - hopefully that addresses the event recorder too ? I kept it in the `default` case switch though - should I remove there too @toVersus  ? Thanks for the review and inputs!

toVersus commented on pkg/kubelet/prober/prober.go:
```diff
@@ -137,13 +144,16 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
 		if err != nil {
+			// Log and record event for Unknown result
+			klog.V(3).InfoS("HTTP-Probe failed to create request", "error", err)
+			pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerProbeWarning, "%s probe failed: %v", probeType, err)
```
If you write an Event here, it will be recorded each time the probe goes to the Unknown state during a retry. But if the issue gets resolved through retries, I don’t think writing it as an Event is necessary. Just keeping it as a log message should be enough for debugging. If retries don’t fix the issue, it'll still be logged as an Event anyway. Same for l.173.

toVersus commented on pkg/kubelet/prober/prober.go:
```diff
@@ -93,29 +93,36 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
+		klog.V(1).InfoS("Probe warning", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "output", output)
+		return results.Success, nil
+
+	case probe.Unknown:
```
If the probe’s state is Unknown, it will always return an error, so I don’t think this part of the code will ever be executed.
https://github.com/kubernetes/kubernetes/blob/c7e6c419868d131127ceccf61f3c077436f1b3bb/pkg/kubelet/prober/prober.go#L157
https://github.com/kubernetes/kubernetes/blob/c7e6c419868d131127ceccf61f3c077436f1b3bb/pkg/kubelet/prober/prober.go#L174
https://github.com/kubernetes/kubernetes/blob/c7e6c419868d131127ceccf61f3c077436f1b3bb/pkg/kubelet/prober/prober.go#L194

If the probe's state is Unknown, it'll go into this error handling.
https://github.com/kubernetes/kubernetes/blob/c7e6c419868d131127ceccf61f3c077436f1b3bb/pkg/kubelet/prober/prober.go#L102-L107

So, I think it'd be better to go with these changes instead.

```diff
diff --git a/pkg/kubelet/prober/prober.go b/pkg/kubelet/prober/prober.go
index 3b771847bf8..08afa16a164 100644
--- a/pkg/kubelet/prober/prober.go
+++ b/pkg/kubelet/prober/prober.go
@@ -98,7 +98,11 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
        }

        result, output, err := pb.runProbeWithRetries(ctx, probeType, probeSpec, pod, status, container, containerID, maxProbeRetries)
-
+       if err != nil && result == probe.Unknown {
+               klog.V(1).ErrorS(err, "Probe unknown", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
+               pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerProbeWarning, "%s probe resulted in unknown state: %v", probeType, err)
+               return results.Failure, err
+       }
        if err != nil {
                // Handle probe error
                klog.V(1).ErrorS(err, "Probe errored", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
@@ -114,10 +118,6 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
                klog.V(1).InfoS("Probe warning", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "output", output)
                return results.Success, nil

-       case probe.Unknown:
-               klog.V(1).InfoS("Probe unknown", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "output", output)
-               return results.Failure, nil
-
        default:
                klog.V(1).InfoS("Probe failed", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "probeResult", result, "output", output)
                pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerUnhealthy, "%s probe failed: %s", probeType, output)
```

bart0sh replied:
This complicates the code and makes an assumption that `probe.Unknown` will always come with non-nil error. We shouldn't assume this. I'm for keeping @jralmaraz approach.

toVersus replied:
> We shouldn't assume this.

Yeah, it’s fine to keep the current switch case as is. But in the current implementation, an error gets returned every time the probe is unknown, so when `runProbeWithRetries` hits the retry limit, it returns an error. Like I mentioned above, this sends it into the previous error handling section and treats everything as a 'Probe errored'. So, if we want to log and write an event as 'Probe unknown', I think the suggested code addition is necessary.

jralmaraz replied:
Thanks for the inputs ! I've tried splitting into `Probe unknown with error` and `Probe unknown without error`. Does it help addressing the concern without adding complication to the code ? 

`	if err != nil && result == probe.Unknown {
		klog.V(1).ErrorS(err, "Probe unknown with error", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
		pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerProbeWarning, "%s probe resulted in unknown state: %v", probeType, err)
		return results.Failure, err
	}

	if result == probe.Unknown {
		klog.V(1).InfoS("Probe unknown without error", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
		return results.Failure, nil
	}
`

bart0sh replied:
I'd prefer the latter to go into `switch` statement below. As for the former, what about including result into the message, so for any result type the error message would be informative enough?

toVersus replied:
> I'd prefer the latter to go into switch statement below.

+1. It's easier to read that way.

```go
	switch result {
	case probe.Success:
		return results.Success, nil

	(...)

	case probe.Unknown:
		klog.V(1).InfoS("Probe unknown without error", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
		return results.Failure, nil

	(...)
	}
```

toVersus commented on pkg/kubelet/prober/prober.go:
```diff
@@ -93,29 +93,36 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
 
 	if probeSpec == nil {
-		klog.InfoS("Probe is nil", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
+		klog.V(3).InfoS("Probe is nil", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
```
I think we should avoid changing the existing log levels in this PR. It’s fine to add new logs, but if we stop writing logs without a good reason, it could cause issues for some users.

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -93,29 +93,36 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
+		klog.V(1).InfoS("Probe unknown", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "output", output)
+		return results.Failure, nil
+
+	default:
```
please use `probe.Failure `explicitly instead of default.

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -98,24 +98,34 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
-			pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerUnhealthy, "%s probe failed: %s", probeType, output)
-		}
+
+	if err != nil && result == probe.Unknown {
```
This only handles err if probe is unknown. I'm not sure this is correct.

toVersus replied:
You did not need to remove the error handling for the probe here.

```go
	result, output, err := pb.runProbeWithRetries(ctx, probeType, probeSpec, pod, status, container, containerID, maxProbeRetries)
	if err != nil && result == probe.Unknown {
		klog.V(1).ErrorS(err, "Probe unknown", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
		pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerProbeWarning, "%s probe resulted in unknown state: %v", probeType, err)
		return results.Failure, err
	}
	if err != nil {
		// Handle probe error
		klog.V(1).ErrorS(err, "Probe errored", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
		pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerUnhealthy, "%s probe errored: %v", probeType, err)
		return results.Failure, err
	}
```

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -98,24 +98,34 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
 	}
```
Code on the lines 95-98 looks unreachable to me as probeSpec can't be nil at this point. I'd suggest to remove it.

bart0sh replied:
My bad. `probeSpec` can still be nil if there if probe is not configured for the container. Please, revert this change. This should fix nil pointer dereference you're debugging.

jralmaraz commented:
thanks a lot @bart0sh and @toVersus . 

I'm trying to troubleshoot a nil pointer dereference I've created with the latest changes 🫠 

```
--- FAIL: TestProbe (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x106fb5918]

goroutine 391 [running]:
testing.tRunner.func1.2({0x1075f12a0, 0x108a36df0})
        /opt/homebrew/Cellar/go/1.22.4/libexec/src/testing/testing.go:1631 +0x2c8
testing.tRunner.func1()
        /opt/homebrew/Cellar/go/1.22.4/libexec/src/testing/testing.go:1634 +0x47c
panic({0x1075f12a0?, 0x108a36df0?})
        /opt/homebrew/Cellar/go/1.22.4/libexec/src/runtime/panic.go:770 +0x124
k8s.io/kubernetes/pkg/kubelet/prober.(*prober).runProbe(_, {_, _}, _, _, _, {{0x0, 0x0}, {0x0, 0x0, ...}, ...}, ...)
        /Users/jalmaraz/dev/git/forks/kubernetes/pkg/kubelet/prober/prober.go:149 +0x58
k8s.io/kubernetes/pkg/kubelet/prober.(*prober).runProbeWithRetries(_, {_, _}, _, _, _, {{0x0, 0x0}, {0x0, 0x0, ...}, ...}, ...)
        /Users/jalmaraz/dev/git/forks/kubernetes/pkg/kubelet/prober/prober.go:140 +0x14c
k8s.io/kubernetes/pkg/kubelet/prober.(*prober).probe(_, {_, _}, _, _, {{0x0, 0x0}, {0x0, 0x0, 0x0}, ...}, ...)
        /Users/jalmaraz/dev/git/forks/kubernetes/pkg/kubelet/prober/prober.go:95 +0x120
k8s.io/kubernetes/pkg/kubelet/prober.TestProbe(0xc0004e0340)
        /Users/jalmaraz/dev/git/forks/kubernetes/pkg/kubelet/prober/prober_test.go:245 +0xc40
testing.tRunner(0xc0004e0340, 0x1078db1e8)
        /opt/homebrew/Cellar/go/1.22.4/libexec/src/testing/testing.go:1689 +0x184
created by testing.(*T).Run in goroutine 1
        /opt/homebrew/Cellar/go/1.22.4/libexec/src/testing/testing.go:1742 +0x5e8
FAIL    k8s.io/kubernetes/pkg/kubelet/prober    9.020s
```

jralmaraz commented:
![image](https://github.com/user-attachments/assets/9ef54c4c-b9cd-4a98-8e70-bac73fc617bc)

Sorry for being so newbie, but that's why I chose a "good first issue" :)

I can see during local debugging that its reaching `pb.runProbe` with a `nil` `p`

bart0sh commented:
@jralmaraz 
> Sorry for being so newbie, but that's why I chose a good first issue :) 

Considering the number of previous PRs mentioned above, marking this as a good first issue was a mistake. However, you went further, so don't give up :)

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -98,24 +98,42 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
+		klog.V(1).ErrorS(err, "Probe unknown with error", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
+		pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerProbeWarning, "%s probe resulted in unknown state: %v", probeType, err)
+		return results.Failure, err
+	}
```
Why err != nil and probe.Unknown are so special so we need to process this particular case separately?

toVersus replied:
You're right, there’s no real need to treat this as a special case. Since an Unknown probe ends up with results.Failure, it makes more sense to handle it as a ContainerUnhealthy event. So, it’d probably be simpler to combine the error handling and include the probe result in the log and event message.

```go
	result, output, err := pb.runProbeWithRetries(ctx, probeType, probeSpec, pod, status, container, containerID, maxProbeRetries)
	if err != nil {
		// Handle probe error
		klog.V(1).ErrorS(err, "Probe errored", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "probeResult", result)
		pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, events.ContainerUnhealthy, "%s probe errored and resulted in %s state: %s", probeType, result, err)
		return results.Failure, err
	}
```

toVersus commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +337,127 @@ func TestNewExecInContainer(t *testing.T) {
+
+}
+
+func TestExecInContainer_Start(t *testing.T) {
```
Why do we need this new test? I think `TestNewExecInContainer` already covered it.

jralmaraz replied:
I was planning to split the exec/setup from the start in case they would need different settings, but, so far, indeed I can just remove it

toVersus commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +337,127 @@ func TestNewExecInContainer(t *testing.T) {
+	assert.Equal(t, runner, prober.runner, "Expected prober runner to match")
+	assert.Equal(t, recorder, prober.recorder, "Expected prober recorder to match")
+
+	assert.NotNil(t, prober.exec, "exec probe initialized")
+
+	assert.NotNil(t, prober.http, "http probe initialized")
+
+	assert.NotNil(t, prober.tcp, "tcp probe initialized")
+
+	assert.NotNil(t, prober.grpc, "grpc probe initialized")
+
```
nit: Since we're checking similar things, I think it's fine to group them together without line breaks.

```suggestion
	assert.NotNil(t, prober.exec, "exec probe initialized")
	assert.NotNil(t, prober.http, "http probe initialized")
	assert.NotNil(t, prober.tcp, "tcp probe initialized")
	assert.NotNil(t, prober.grpc, "grpc probe initialized")
```

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -98,24 +98,36 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
+		return results.Failure, nil
+
+	default:
+		klog.V(1).InfoS("Unhandled probe result", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "probeResult", result)
```
nit: Unhandled probe result -> Unsupported probe result

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -137,13 +149,15 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
 	switch {
 	case p.Exec != nil:
-		klog.V(4).InfoS("Exec-Probe runProbe", "pod", klog.KObj(pod), "containerName", container.Name, "execCommand", p.Exec.Command)
+		klog.V(3).InfoS("Exec-Probe runProbe", "pod", klog.KObj(pod), "containerName", container.Name, "execCommand", p.Exec.Command)
```
@jralmaraz Can you explain the idea behind changing log level here and below? From the first look it's not related to the PR that claims to improve handling of the `probe.Warning` result.

jralmaraz replied:
I think I got them mixed up, more of a mistake on my end. Moving it back

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +337,102 @@ func TestNewExecInContainer(t *testing.T) {
+	}
+
+	container := pod.Spec.Containers[0]
+	recorder := record.NewFakeRecorder(10)
```
Why 10? what happens if test cases will create more than 10 events?

jralmaraz replied:
The [NewFakeRecorder](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/record/fake.go#L80) needs a buffer size. so, I am changing that to the amount of expected events in the test cases + 1 and hopefully that covers this in a more "dynamic" way ?

jralmaraz commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +337,102 @@ func TestNewExecInContainer(t *testing.T) {
+	for _, tc := range testCases {
+
+		// Create a buffered channel with capacity of 2 - trying to ensure that even if events are not immediately consumed, the event recorder won't block.
+		recorder.Events = make(chan string, 2)
```
indeed - I have removed it

jralmaraz commented:
/test all

jralmaraz commented:
/retest

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +337,106 @@ func TestNewExecInContainer(t *testing.T) {
+			pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, "ContainerProbeWarning", "%s probe warning: %s", tc.probeType, output)
+			pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, "ContainerProbeWarning", "Unknown %s probe status: %s", tc.probeType, tc.result)
+
+			var events []string
+			for i := 0; i < len(tc.expected); i++ {
+				select {
+				case event := <-fakeRecorder.Events:
+					events = append(events, event)
+				default:
+					assert.Failf(t, "missing events", "expected %d events, but got fewer", len(tc.expected))
+					return
+				}
+			}
+
+			select {
+			case extraEvent := <-fakeRecorder.Events:
+				assert.Failf(t, "unexpected event", "unexpected extra event: %s", extraEvent)
+			default:
+				// No extra event, proceed with validation
+			}
+
+			assert.Equal(t, tc.expected, events, "unexpected events")
```
```suggestion
                         assert.Equal(t, len(tc.expected), len(fakeRecorder.Events), "unexpected number of events")
                         for _, expected := range tc.expected {
                                 assert.Equal(t, expected, <-fakeRecorder.Events)
                         }
```

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +337,106 @@ func TestNewExecInContainer(t *testing.T) {
+	}
+
+	for i, tc := range testCases {
+		t.Run(fmt.Sprintf("%d-%s", i, tc.probeType), func(t *testing.T) {
```
Can you either add a name/description fields to the test case structure and output it here or at least add `result` to this generated test case name?

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +337,106 @@ func TestNewExecInContainer(t *testing.T) {
+			bufferSize := len(tc.expected) + 1
+			fakeRecorder := record.NewFakeRecorder(bufferSize)
+
+			pb := &prober{
```
Moving `proper` creation into the loop as another good implication: it isolates test cases from each other. 
+1

jralmaraz commented:
/retest

jralmaraz commented:
/retest

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +337,102 @@ func TestNewExecInContainer(t *testing.T) {
+			pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, "ContainerProbeWarning", "%s probe warning: %s", tc.probeType, output)
+			pb.recordContainerEvent(pod, &container, v1.EventTypeWarning, "ContainerProbeWarning", "Unknown %s probe status: %s", tc.probeType, tc.result)
+
+			// Check if the expected number of events are available
+			if len(fakeRecorder.Events) != len(tc.expected) {
+				t.Errorf("Unexpected number of events. Expected: %d, Actual: %d", len(tc.expected), len(fakeRecorder.Events))
+				return
+			}
```
Isn't it the same as the next line?
```suggestion
```

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +337,102 @@ func TestNewExecInContainer(t *testing.T) {
+			for _, expected := range tc.expected {
+				assert.Equal(t, expected, <-fakeRecorder.Events)
+			}
+			assert.Equal(t, probe.Unknown, tc.result)
```
This doesn't seem to be correct. It validates test case data, but is it really needed here?

bart0sh replied:
I'm not sure about that. This code doesn't test events and only tests failures as far as I can see. How about revert the latest change and apply review suggestions?

jralmaraz replied:
Sounds good @bart0sh - thanks for your patience on this. I've tried to add the assertion to check the expected number of events and added a content check comparing event and expected in test case.

bart0sh commented on pkg/kubelet/prober/prober_test.go:
```diff
@@ -333,3 +337,104 @@ func TestNewExecInContainer(t *testing.T) {
+
+			assert.Equal(t, len(tc.expected), len(fakeRecorder.Events), "unexpected number of events")
+
+			for _, expected := range tc.expected {
+				found := false
+				for i := 0; i < len(fakeRecorder.Events); i++ {
+					actualEvent := <-fakeRecorder.Events
+					if strings.Contains(actualEvent, expected) {
+						found = true
+						break
+					}
+				}
+				require.True(t, found, "expected message '%s' not found in events", expected)
+			}
```
How this is different from what [was suggested before](https://github.com/kubernetes/kubernetes/pull/125901#discussion_r1722845825)?

jralmaraz replied:
You are right, I don't think there's much of a difference, rolled back.

jralmaraz replied:
Let me know if its ok to resolve this one.

bart0sh replied:
yes, please resolve.

bart0sh commented on pkg/kubelet/prober/prober.go:
```diff
@@ -98,24 +98,36 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
+
+	default:
+		klog.V(1).InfoS("Unsupported probe result", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name, "probeResult", result)
+		return results.Failure, nil
```
@jralmaraz It would be nice to cover this new code by unit tests.

jralmaraz replied:
I tried to add one, but I am not sure how to include a check to verify the Info Log content. If I would need to create a buffer and redirect it or something else. For now I am going with an assertion for the default returned error.

bart0sh replied:
Checking return value should be enough, I believe.

jralmaraz replied:
Let me know if the unit test I've pushed is ok and I'll resolve this one. Thanks!

bart0sh replied:
Can `TestProbe` be extended instead of adding new test?

jralmaraz replied:
I've just tried that by adding a flag and had to rename the `probeType` variable so I could set an invalid one, as it was conflicting. Is that ok ? Thanks!

bart0sh replied:
Thanks, it looks ok to me. I'd remove exact error check though:
```
				if test.unsupported {
					require.Contains(t, err.Error(), "unknown probe type", "[%s] Expected 'unknown probe type' error but got %v", testID, err)
				}
```

jralmaraz replied:
Thanks. Let me know if I can resolve this one.

bart0sh commented:
@jralmaraz Please, squash commits into one.

jralmaraz commented:
> @jralmaraz Please, squash commits into one.

thanks @bart0sh , just pushed an update. Let me know if there's anything else I should address. Cheers!

bart0sh commented:
@jralmaraz last step - can you shorten your commit message and title. It's too verbose.
The rest looks good to me. Thank you for your contributions!

bart0sh commented:
retitle Report event for the cases when probe returned Unknown result

bart0sh commented:
/retitle Report event for the cases when probe returned Unknown result

jralmaraz commented:
> @jralmaraz last step - can you shorten your commit message and title? It's too verbose. The rest looks good to me. Thank you for your contributions!

Sounds good @bart0sh , thanks for all your help and support. Let me know if there's anything else on my end. Cheers!

pacoxu commented on pkg/kubelet/prober/prober.go:
```diff
@@ -98,24 +98,36 @@ func (pb *prober) probe(ctx context.Context, probeType probeType, pod *v1.Pod, s
+
+	switch result {
+	case probe.Success:
+		return results.Success, nil
```
https://github.com/kubernetes/kubernetes/blob/a4c262bc8cdf4002b736f11ae847e1e8fcfef1f2/pkg/kubelet/prober/prober.go#L116C1-L117C1

The succeed log should not be removed, IIUC.

jralmaraz replied:
should it remain .V(3) level as well @pacoxu  ?

pacoxu replied:
Yes, if we don't have a reason to change it?

jralmaraz replied:
Let me know if it looks good and I can resolve this @pacoxu  - thanks for your input

jralmaraz commented:
hi @bart0sh  - do you mind having a look again after the latest change ? Thanks!

bart0sh commented:
/lgtm

bart0sh commented:
@SergeyKanzhelev @mrunalp PTAL. Considering [amount of PRs trying to fix the same issue](https://github.com/kubernetes/kubernetes/pull/125901#issuecomment-2249883983), it would be great if we merge this.

dims commented:
@bart0sh the change looks good to me as well, if it does not merge in another week, please ping me. This has been waiting too long unfortunately!

bart0sh commented:
@dims pinging as agreed.
